include/nbl/video/IDescriptorPool.h

```diff
@@ -29,12 +36,99 @@ class NBL_API IDescriptorPool : public core::IReferenceCounted, public IBackendO
+                    }
+                    else
+                    {
+                        m_linearAllocators[i] = core::LinearAddressAllocator<uint32_t>(nullptr, 0u, 0u, 1u, m_maxDescriptorCount[i]);
+                    }
+                }
+            }
+
+            // For (possibly) mutable samplers.
+            if (m_flags & ECF_FREE_DESCRIPTOR_SET_BIT)
+            {
+                m_generalAllocatorReservedSpace[asset::EDT_COUNT] = std::make_unique<uint8_t[]>(core::GeneralpurposeAddressAllocator<uint32_t>::reserved_size(1u, m_maxDescriptorCount[asset::EDT_COMBINED_IMAGE_SAMPLER], 1u));
+                m_generalAllocators[asset::EDT_COUNT] = core::GeneralpurposeAddressAllocator<uint32_t>(m_generalAllocatorReservedSpace[asset::EDT_COUNT].get(), 0u, 0u, 1u, m_maxDescriptorCount[asset::EDT_COMBINED_IMAGE_SAMPLER], 1u);
+            }
+            else
+            {
+                m_linearAllocators[asset::EDT_COUNT] = core::LinearAddressAllocator<uint32_t>(nullptr, 0u, 0u, 1u, m_maxDescriptorCount[asset::EDT_COMBINED_IMAGE_SAMPLER]);
+            }
 
-        uint32_t getCapacity() const { return m_maxSets; }
+            // Initialize the storages.
+            m_textureStorage = std::make_unique<core::StorageTrivializer<core::smart_refctd_ptr<video::IGPUImageView>>[]>(m_maxDescriptorCount[asset::EDT_COMBINED_IMAGE_SAMPLER]);
+            m_mutableSamplerStorage = std::make_unique<core::StorageTrivializer<core::smart_refctd_ptr<video::IGPUSampler>>[]>(m_maxDescriptorCount[asset::EDT_COMBINED_IMAGE_SAMPLER]);
+            m_storageImageStorage = std::make_unique<core::StorageTrivializer<core::smart_refctd_ptr<IGPUImageView>>[]>(m_maxDescriptorCount[asset::EDT_STORAGE_IMAGE] + m_maxDescriptorCount[asset::EDT_INPUT_ATTACHMENT]);
+            m_UBO_SSBOStorage = std::make_unique<core::StorageTrivializer<core::smart_refctd_ptr<IGPUBuffer>>[]>(m_maxDescriptorCount[asset::EDT_UNIFORM_BUFFER] + m_maxDescriptorCount[asset::EDT_STORAGE_BUFFER] + m_maxDescriptorCount[asset::EDT_UNIFORM_BUFFER_DYNAMIC] + m_maxDescriptorCount[asset::EDT_STORAGE_BUFFER_DYNAMIC]);
+            m_UTB_STBStorage = std::make_unique<core::StorageTrivializer<core::smart_refctd_ptr<IGPUBufferView>>[]>(m_maxDescriptorCount[asset::EDT_UNIFORM_TEXEL_BUFFER] + m_maxDescriptorCount[asset::EDT_STORAGE_BUFFER_DYNAMIC]);
+            m_accelerationStructureStorage = std::make_unique<core::StorageTrivializer<core::smart_refctd_ptr<IGPUAccelerationStructure>>[]>(m_maxDescriptorCount[asset::EDT_ACCELERATION_STRUCTURE]);
+        }
+
+        ~IDescriptorPool() {}
+
+        // Returns the offset into the pool's descriptor storage. These offsets will be combined
+        // later with base memory addresses to get the actual memory adress where we put the core::smart_refctd_ptr<const IDescriptor>.
+        SDescriptorOffsets allocateDescriptors(const IGPUDescriptorSetLayout* layout);
```

Comment:
this should not be public, you should name it `allocateDescriptorOffsets` and make it protected

all of these methods (including the protected virtual `impl`) https://github.com/Devsh-Graphics-Programming/Nabla/blob/descriptor_lifetime_tracking/include/nbl/video/ILogicalDevice.h#L235-L292 should move from `ILogicalDevice` to here

Reply:
As per [this comment](https://github.com/Devsh-Graphics-Programming/Nabla/pull/345#discussion_r983874583), `allocateDescriptors` is moved to `IDescriptorSet` as a pure `virtual` method. Do we still rename it to `allocateDescriptorOffsets`?

Also due to the above linked comment, the implementations (`IGPUDescriptorSet::allocateDescriptors` and `ICPUDescriptorSet::allocateDescriptors`) are both be made `private`.

Reply:
Resolved, except the renaming and making it `protected` part because it is in direct contradiction with the comment I linked above.

Reply:
> As per https://github.com/Devsh-Graphics-Programming/Nabla/pull/345#discussion_r983874583, allocateDescriptors is moved to IDescriptorSet as a pure virtual method. Do we still rename it to allocateDescriptorOffsets?

renaming is a good idea.

Reply:
Resolved.